### ISSUE TITLE
Username Enumeration

### ISSUE THREAD

**Issue by senthilkumardhanapal2** [Azure/api-management-developer-portal#1399]
The Developer portal forgot password functionality that allowed an attacker to determine if a given username was valid. Specifically, when accessing the signup and reset password pages, the application's response differed depending on if a valid or unknown username was provided. 

```
usecase 1:
When correct username was provided , got a successful response and asked to follow the instructions

usecase 2:
When incorrect username was provided , no message was displayed , only page was refreeshed. end user do not have an idea where input in invalid or any other error.
```

```
Affected Scope :
https://prodprhtzinterop.developer.healthtranzformdev.com/reset-password
```
```
Impact : 
An attacker can generate a list of probable usernames and easily determine which are valid. The valid usernames can then be used in subsequent attacks, such as password brute force.
```

```
Recommendations : 
Update the APIM dev portal forgotpassword function to display generic error messages for both valid and invalid usernames. 
```

**Comment by mikebudzynski** [comment]
@senthilkumardhanapal2, could you explain why this is an issue? We're also open to hearing your suggestions for the improvements. Many online services, like GMail or GitHub, have the endpoints to check if the username already exists (for example, this logic is used in the sign-up flow to check the account name availability). The managed portal employs Captcha to greatly reduce the risk of abuse by bots.

To further mitigate the issue, we can tighten the throttling limits on the sign-in, sign-up, reset-password, and forgot-password endpoints.

**Comment by senthilkumardhanapal2** [comment]
Mike, issue is when we use APIM management for users (Username and Password), it doesn't return any error page. We don't even know request was failure r success. To me it looks like CSS / error component issue. Same behavior observed during signup as well.

**Comment by mikebudzynski** [comment]
> Mike, issue is when we use APIM management for users (Username and Password), it doesn't return any error page. We don't even know request was failure r success. To me it looks like CSS / error component issue. Same behavior observed during signup as well.

@senthilkumardhanapal2, sorry, I don't understand your response. Could you elaborate?

I explained above why we don't think the reported "issue" is actually a security issue and can be used as an attack vector. In short:

- It's common for online services to surface the information whether the username is already taken or not
- API Management uses Captcha for protection against bots
- API Management uses throttling for protection against bots

I understand that this:

> When incorrect username was provided , no message was displayed , only page was refreeshed. end user do not have an idea where input in invalid or any other error.

is a usability issue that we can address, but it is not what the original post is referring to as the actual problem.

**Comment by papadakis-cts** [comment]
Hi Mike, 

This came from a third-party pen test we had executed on a solution that uses APIM and the dev portal.  The objection noted is one can use this is to determine if the user account exists.  The recommended solution by the vendor is to have an error message that is more generic not giving the user feedback on if the account exists or not.  In theory, the current message gives a potential hacker information on the existence of the account.

I would be happy to discuss this more, please reach out to me as needed (elliot.papadakis@cognizant.com)

Thanks,

Elliot

**Comment by madhancock** [comment]
> Hi Mike,
> 
> This came from a third-party pen test we had executed on a solution that uses APIM and the dev portal. The objection noted is one can use this is to determine if the user account exists. The recommended solution by the vendor is to have an error message that is more generic not giving the user feedback on if the account exists or not. In theory, the current message gives a potential hacker information on the existence of the account.
> 
> I would be happy to discuss this more, please reach out to me as needed ([elliot.papadakis@cognizant.com](mailto:elliot.papadakis@cognizant.com))
> 
> Thanks,
> 
> Elliot

Pen testers often report this but the fix isn't as simple as making the message more generic. Attackers can use response times or backend HTTP codes to determine if a user exists or not.

Solving these still don't actually soft the root issue if your application has any form of signup form. Instead they make your application less user friendly, but an attacker can easily test if an account exists by using your signup form - any error on email address is likely to indicate that the account exists, regardless of the message.

The correct solution would be to use anti-automation techniques such as Captcha and WAF.

Both of these are implemented in API Management already and would merit only an "info" level mention on a application security test. Only if a tester can demonstrate that the implemented mitigations are ineffective at preventing effective enumeration would it be of concern.

**Comment by mikebudzynski** [comment]
@papadakis-cts, I agree with @madhancock's response above; see my previous comments in this thread. The sign-up endpoint already informs the user whether a username is taken - it's a common practice in web development and I wouldn't consider it a security vulnerability. We protect against scraping or automated attacks with Captcha and throttling. If the tests find that these protection levels are too permissive, we'll gladly look into tightening the security measures.

**Comment by azaslonov** [comment]
Looks like we've come to an agreement. Hence, closing it.
